rpc/v9/subscriptions.go

```diff
@@ -204,50 +218,268 @@ func (h *Handler) SubscribeEvents(
-			return nil
 		},
 		onPendingData: func(ctx context.Context, id string, _ *subscription, pending core.PendingData) error {
-			// We should only return `ACCEPTED_ON_L2` events in this stream.
-			if pending == nil || pending.Variant() != core.PendingBlockVariant {
-				return nil
+			var blockFinalityStatus TxnFinalityStatusWithoutL1
+			switch v := pending.Variant(); v {
+			case core.PendingBlockVariant:
+				blockFinalityStatus = TxnFinalityStatusWithoutL1(TxnAcceptedOnL2)
+			case core.PreConfirmedBlockVariant:
+				if *finalityStatus != TxnFinalityStatusWithoutL1(TxnPreConfirmed) {
+					return nil
+				}
+				blockFinalityStatus = TxnFinalityStatusWithoutL1(TxnPreConfirmed)
+			default:
+				return fmt.Errorf("unknown pending variant %v", v)
 			}
 
-			return h.processEvents(
-				ctx, w, id, nextBlock, nextBlock, fromAddr, keys, eventsPreviouslySent,
+			return processBlockEvents(
+				ctx,
+				w,
+				id,
+				pending.GetBlock(),
+				fromAddr,
+				&eventMatcher,
+				eventsPreviouslySent,
+				blockFinalityStatus,
 			)
 		},
 	}
 	return h.subscribe(ctx, w, subscriber)
 }
 
+// processEvents queries database for events and stream filtered events.
+func (h *Handler) processEvents(
+	ctx context.Context,
+	w jsonrpc.Conn,
+	id string,
+	from, to *BlockID,
+	fromAddr *felt.Felt,
+	keys [][]felt.Felt,
+	eventsPreviouslySent map[SentEvent]TxnFinalityStatusWithoutL1,
+	height uint64,
+	finalityStatus TxnFinalityStatusWithoutL1,
+) error {
+	filter, err := h.bcReader.EventFilter(fromAddr, keys, h.PendingBlock)
+	if err != nil {
+		return err
+	}
+
+	defer h.callAndLogErr(filter.Close, "error closing event filter in events subscription")
+
+	err = setEventFilterRange(filter, from, to, height)
+	if err != nil {
+		return err
+	}
+
+	filteredEvents, cToken, err := filter.Events(nil, subscribeEventsChunkSize)
+	if err != nil {
+		return err
+	}
+
+	err = sendEvents(ctx, w, filteredEvents, eventsPreviouslySent, id, finalityStatus)
+	if err != nil {
+		return err
+	}
+
+	for cToken != nil {
+		filteredEvents, cToken, err = filter.Events(cToken, subscribeEventsChunkSize)
+		if err != nil {
+			return err
+		}
+
+		err = sendEvents(ctx, w, filteredEvents, eventsPreviouslySent, id, finalityStatus)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// processBlockEvents, extract events from block and stream filtered events.
+func processBlockEvents(
+	ctx context.Context,
+	w jsonrpc.Conn,
+	id string,
+	block *core.Block,
+	fromAddr *felt.Felt,
+	eventMatcher *blockchain.EventMatcher,
+	eventsPreviouslySent map[SentEvent]TxnFinalityStatusWithoutL1,
+	finalityStatus TxnFinalityStatusWithoutL1,
+) error {
+	if isMatch := eventMatcher.TestBloom(block.EventsBloom); !isMatch {
+		return nil
+	}
+
+	var blockNumber *uint64
+	// if header.Hash == nil and parentHash != nil it's a pending block
+	// if header.Hash == nil and parentHash == nil it's a pre_confirmed block
+	if block.Hash != nil || block.ParentHash == nil {
```

Comment:
nit:
```suggestion
	if isNotPending := block.Hash != nil || block.ParentHash == nil; isNotPending {
```
feel free to ignore